### ISSUE TITLE
Is there a way to know parameters used in a model saved?

### ISSUE THREAD

**Issue by BlindApe** [dmlc/xgboost#196]
mainly, the base_score. 

I'm trying to reproduce prediction of a model from txt dump, so:

1) I score each booster
2) sum the scores
3) add the base_score (in margin value scale)
4) transform to probabilities

Is this procedure right?

**Comment by pommedeterresautee** [comment]
FYI there is a function `xgb.model.dt.tree` which parse the txt model and convert it to a data.table. It will be easier to play with it in this format.

**Comment by pommedeterresautee** [comment]
BTW if you can post your code somewhere when it is ready, I would be interested to look at it. 
It may give some interesting intuition in a feature importance analysis for instance.

**Comment by BlindApe** [comment]
Thank you.
Really I haven't the code. It is developed by people tha want to integrate the prediction of a model in other language and ask to me this issue. So convert to dt in this case isn't useful.

**Comment by pommedeterresautee** [comment]
May be a conversion of the model to pmml format would be an interesting feature for that purpose!

**Comment by pommedeterresautee** [comment]
@tqchen do you think pmml is an interesting feature? If yes, we may ask to people from http://cran.r-project.org/web/packages/pmml/ what they need to make xgboost compliant with their package (XML is not funny I prefer to let the pro do the dirty work).

**Comment by tqchen** [comment]
I think dump into some common format was good, but I do not know what could be a common format for trees. I guess the text dump contains most of the information you need to re-created the result(need base_marginscore).  Maybe json is another format that worth considering

**Comment by pommedeterresautee** [comment]
Pmml supports trees:
http://www.dmg.org/v3-2/TreeModel.html

And pmml R package supports Adaboost already (http://cran.r-project.org/web/packages/pmml/pmml.pdf)

**Comment by chrissly31415** [comment]
I have a related problem. I used the xgb.model.dt.tree function to get the structure of a trained tree, but its not really clear to me from the documentation how to interpret it and if the information there is sufficient to make a full prediction.

-For regression I get sometimes a splitvalue eqal to "NA" (and its not a leaf)?
-Does the quality feature for a leaf correspond to the final predicted value? If not where would I get that final prediction?
-Is there somewhere a base value/average to start the prediction i.e. 0th iteration/average like in gbm?

P.S. I believe the pmml-format  will be very verbose resulting in quite huge text files even for moderate sized tree ensembles.

**Comment by pommedeterresautee** [comment]
Hi @chrissly31415,

I wrote this function to produce a data.table that I can use for feature importance analysis and drawing the trees (and few other things). It is basically the result of the parsing of the text model dump in a format which makes it easy to play with. So your real question is more can you make prediction with the information included in the text model dump?

The main information you may need to compute is the weight of the vote of each tree. I would say it depends in some way of the gain of the splits for each branch but I may be wrong on that. I don't think you will find direct answer in the slides of @tqchen but it s certainly a good place to begin with. Source code may help too.

The NA you are refering to are for the leafs. If you post an example (data.table + dump text model) I can help you in the understanding and improve the function documentation if required.

Kind regards,
Michaël

**Comment by chrissly31415** [comment]
Hi Michael,

many thanks for the prompt reply. I probably have answered the question already by myself looking at the text dump (sorry, I was not aware of that function before) .
-Regarding the first point, it seems to be a bug, looks like a negative split points produce a "NA" in the xgb.model.dt.tree function i.e. it should be:
splitBranch <- str_extract(branch, "<-?\\d*\\.*\\d*\\]") %>% str_replace("<", "") %>% str_replace("\\]", "")
in the xgb.model.dt.tree.R

-the final leaf value/prediction (leaf=xxx) is shown as "quality" in the datatable object

Best regards,
Christoph

**Comment by chrissly31415** [comment]
Hi,

finally I was able to make prediction for a regression based on the xgb.model.dt.tree or the model dump.
By the way, iteration 0 starts with y=0.5 (for regression with gbm package its the average of the y,training data, but I guess it does not matter).
Concerning the xgb.model.dt.tree I submitted a pull request related to the parsing of numbers (negative reals and scientific notation were not parsed and NA was created as split value).
Still, there is possibly a bug in the R predict function, whenever a sample has a feature with the value of exactly = 0, it might take the wrong turn within the tree.
A workaround was to use something like data[tdata == 0] <- 1E-15 before prediction. 

Best regards,
Christoph
